docs/concepts/custom-ui-headless-api.mdx

```diff
@@ -0,0 +1,662 @@
+already initialized the flow and redirected to the UI. In such a case, the flow ID can be extracted from the URL `?flow=` query
+parameter.
+
+Here are some cases where the flow would already exist:
+
+1. Ory has initialized the flow and redirects to the UI (e.g. `/login?flow=<uuidv4>`)
+2. The request was completed by the user, however, an error occurred, and we need to retrieve the flow data to display the error
+3. The UI has 'stored' the flow ID in the URL so that page refreshes can retrieve the already existing flow data
+
+:::note
+
+The flow ID has an expiration time, so it is important to check the response status code when retrieving the flow data using an
+existing flow ID. It is up to your application to handle this error case and create a new flow.
+
+:::
+
+## Submitting a flow
+
+The flow always has to be initialized before any user data (e.g. email and password) can be submitted. Depending on the type of
+application (e.g. browser or native) the initialized flow will set a CSRF cookie and return a CSRF token in the response body. It
+is important in **browser based applications** to take note of the CSRF tokens since it is required when submitting the user data.
+
+:::warning
+
+Never use API flows to implement Browser applications! Using API flows in Single-Page-Apps as well as server-side apps opens up
+several potential attack vectors, including Login and other CSRF attacks.
+
+:::
+
+## Login flow
+
+Login flows are used to authenticate users. When the login is successful, the request will respond with a `200 OK` HTTP status
+code and a JSON body containing the user's identity information. When the request fails, it responds with the appropriate HTTP
+status and error messages.
+
+### Create Login flow
+
+In browser and native applications, additional options can be specified when creating a login flow such as `?refresh=true` and
+`?aal=<level>`. For more information on these options, take a look at the [Self-service flow](../kratos/self-service)
+documentation.
+
+````mdx-code-block
+<Tabs groupId="sdk-examples">
+
+<TabItem value="go" label="Go" default>
+
+```mdx-code-block
+import createLoginApi from '!!raw-loader!../../code-examples/sdk/go/selfservice/create-login.go'
+
+<CodeBlock language="go">{createLoginApi}</CodeBlock>
+```
+
+</TabItem>
+<TabItem value="typescript" label="TypeScript">
+
+```mdx-code-block
+import createTsLoginApi from '!!raw-loader!../../code-examples/sdk/typescript/src/selfservice/login/create-login-native.ts'
+
+<CodeBlock language="ts">{createTsLoginApi}</CodeBlock>
+```
+
+</TabItem>
+
+<TabItem value="react" label="React">
+
+```mdx-code-block
+import createReactLoginApi from '!!raw-loader!../../code-examples/sdk/typescript/src/selfservice/login/create-login-react.tsx'
+
+<CodeBlock language="ts">{createReactLoginApi}</CodeBlock>
+```
+
+</TabItem>
+
+
+</Tabs>
+````
+
+#### Two-Factor authentication
```

Comment:
We should add to this section that an login flow is also used to upgrade an `aal1` to an `aal2` session.